File: backend/hanihome-au-api/src/main/java/com/hanihome/hanihome_au_api/domain/entity/Property.java
package com.hanihome.hanihome_au_api.domain.entity;

import com.hanihome.hanihome_au_api.domain.enums.PropertyStatus;
import com.hanihome.hanihome_au_api.domain.enums.PropertyType;
import com.hanihome.hanihome_au_api.domain.enums.RentalType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "properties", indexes = {
    @Index(name = "idx_property_type", columnList = "propertyType"),
    @Index(name = "idx_rental_type", columnList = "rentalType"),
    @Index(name = "idx_status", columnList = "status"),
    @Index(name = "idx_monthly_rent", columnList = "monthlyRent"),
    @Index(name = "idx_deposit", columnList = "deposit"),
    @Index(name = "idx_address", columnList = "address"),
    @Index(name = "idx_available_date", columnList = "availableDate"),
    @Index(name = "idx_created_date", columnList = "createdDate"),
    @Index(name = "idx_landlord_id", columnList = "landlordId")
})
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@EntityListeners(AuditingEntityListener.class)
public class Property {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 200)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false, length = 500)
    private String address;

    @Column(length = 100)
    private String detailAddress;

    @Column(length = 10)
    private String zipCode;

    @Column(length = 50)
    private String city;

    @Column(length = 50)
    private String district;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PropertyType propertyType;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private RentalType rentalType;

    @Column(precision = 12, scale = 0)
    private BigDecimal deposit;

    @Column(precision = 10, scale = 0)
    private BigDecimal monthlyRent;

    @Column(precision = 12, scale = 0)
    private BigDecimal maintenanceFee;

    @Column(precision = 8, scale = 2)
    private BigDecimal area;

    @Column
    private Integer rooms;

    @Column
    private Integer bathrooms;

    @Column
    private Integer floor;

    @Column
    private Integer totalFloors;

    @Column
    private LocalDate availableDate;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private PropertyStatus status = PropertyStatus.PENDING_APPROVAL;

    @Column(nullable = false)
    private Long landlordId;

    @Column
    private Long agentId;

    @ElementCollection
    @CollectionTable(name = "property_options", joinColumns = @JoinColumn(name = "property_id"))
    @Column(name = "option_name")
    @Builder.Default
    private List<String> options = new ArrayList<>();

    @ElementCollection
    @CollectionTable(name = "property_images", joinColumns = @JoinColumn(name = "property_id"))
    @Column(name = "image_url", length = 500)
    @OrderColumn(name = "image_order")
    @Builder.Default
    private List<String> imageUrls = new ArrayList<>();

    @Column(precision = 10, scale = 7)
    private BigDecimal latitude;

    @Column(precision = 10, scale = 7)
    private BigDecimal longitude;

    @Column
    private Boolean parkingAvailable;

    @Column
    private Boolean petAllowed;

    @Column
    private Boolean furnished;

    @Column
    private Boolean shortTermAvailable;

    @Column(columnDefinition = "TEXT")
    private String adminNotes;

    @Column
    private LocalDateTime approvedAt;

    @Column
    private Long approvedBy;

    @CreatedDate
    @Column(nullable = false, updatable = false)
    private LocalDateTime createdDate;

    @LastModifiedDate
    @Column(nullable = false)
    private LocalDateTime modifiedDate;

    @Version
    private Long version;

    public void approve(Long approvedBy) {
        this.status = PropertyStatus.ACTIVE;
        this.approvedAt = LocalDateTime.now();
        this.approvedBy = approvedBy;
    }

    public void reject() {
        this.status = PropertyStatus.REJECTED;
    }

    public void suspend() {
        this.status = PropertyStatus.SUSPENDED;
    }

    public void complete() {
        this.status = PropertyStatus.COMPLETED;
    }

    public void activate() {
        this.status = PropertyStatus.ACTIVE;
    }

    public void deactivate() {
        this.status = PropertyStatus.INACTIVE;
    }

    public boolean isActive() {
        return status == PropertyStatus.ACTIVE;
    }

    public boolean isPendingApproval() {
        return status == PropertyStatus.PENDING_APPROVAL;
    }

    public boolean isCompleted() {
        return status == PropertyStatus.COMPLETED;
    }
}

File: frontend/hanihome-au/src/lib/types/property.ts
// Property related types
export enum PropertyType {
  APARTMENT = 'APARTMENT',
  VILLA = 'VILLA',
  STUDIO = 'STUDIO',
  TWO_ROOM = 'TWO_ROOM',
  THREE_ROOM = 'THREE_ROOM',
  OFFICETEL = 'OFFICETEL',
  HOUSE = 'HOUSE',
}

export enum RentalType {
  MONTHLY = 'MONTHLY',
  JEONSE = 'JEONSE',
  SALE = 'SALE',
}

export enum PropertyStatus {
  PENDING_APPROVAL = 'PENDING_APPROVAL',
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
  SUSPENDED = 'SUSPENDED',
  REJECTED = 'REJECTED',
  COMPLETED = 'COMPLETED',
}

export const PropertyTypeDisplayNames = {
  [PropertyType.APARTMENT]: '아파트',
  [PropertyType.VILLA]: '빌라',
  [PropertyType.STUDIO]: '원룸',
  [PropertyType.TWO_ROOM]: '투룸',
  [PropertyType.THREE_ROOM]: '쓰리룸',
  [PropertyType.OFFICETEL]: '오피스텔',
  [PropertyType.HOUSE]: '단독주택',
};

export const RentalTypeDisplayNames = {
  [RentalType.MONTHLY]: '월세',
  [RentalType.JEONSE]: '전세',
  [RentalType.SALE]: '매매',
};

export interface PropertyFormData {
  // Basic Information
  title: string;
  description: string;
  propertyType: PropertyType;
  rentalType: RentalType;

  // Location Information
  address: string;
  detailAddress?: string;
  zipCode?: string;
  city?: string;
  district?: string;
  latitude?: number;
  longitude?: number;

  // Price Information
  deposit?: number;
  monthlyRent?: number;
  maintenanceFee?: number;

  // Property Details
  area?: number;
  rooms?: number;
  bathrooms?: number;
  floor?: number;
  totalFloors?: number;
  availableDate?: string;

  // Features
  parkingAvailable?: boolean;
  petAllowed?: boolean;
  furnished?: boolean;
  shortTermAvailable?: boolean;

  // Options and Images
  options: string[];
  imageUrls: string[];
}

export interface PropertyFormStep {
  id: string;
  title: string;
  description: string;
  fields: (keyof PropertyFormData)[];
}

export const PROPERTY_FORM_STEPS: PropertyFormStep[] = [
  {
    id: 'basic',
    title: '기본 정보',
    description: '매물의 기본 정보를 입력해주세요.',
    fields: ['title', 'description', 'propertyType', 'rentalType'],
  },
  {
    id: 'details',
    title: '상세 정보',
    description: '매물의 상세 정보를 입력해주세요.',
    fields: [
      'address',
      'detailAddress',
      'area',
      'rooms',
      'bathrooms',
      'floor',
      'totalFloors',
      'availableDate',
      'deposit',
      'monthlyRent',
      'maintenanceFee',
    ],
  },
  {
    id: 'options',
    title: '옵션 선택',
    description: '매물의 옵션과 특징을 선택해주세요.',
    fields: ['parkingAvailable', 'petAllowed', 'furnished', 'shortTermAvailable', 'options'],
  },
  {
    id: 'images',
    title: '이미지 업로드',
    description: '매물 사진을 업로드해주세요.',
    fields: ['imageUrls'],
  },
];

// Property type specific configuration
export const PROPERTY_TYPE_CONFIG = {
  [PropertyType.STUDIO]: {
    showRooms: false,
    showBathrooms: true,
    defaultRooms: 1,
    requiredFields: ['area', 'deposit', 'monthlyRent'],
    helpText: '원룸의 경우 방 개수는 자동으로 1개로 설정됩니다.',
  },
  [PropertyType.TWO_ROOM]: {
    showRooms: true,
    showBathrooms: true,
    defaultRooms: 2,
    requiredFields: ['area', 'rooms', 'deposit', 'monthlyRent'],
    helpText: '투룸은 일반적으로 방 2개, 거실 1개로 구성됩니다.',
  },
  [PropertyType.THREE_ROOM]: {
    showRooms: true,
    showBathrooms: true,
    defaultRooms: 3,
    requiredFields: ['area', 'rooms', 'bathrooms', 'deposit', 'monthlyRent'],
    helpText: '쓰리룸은 방 3개와 거실, 주방으로 구성됩니다.',
  },
  [PropertyType.APARTMENT]: {
    showRooms: true,
    showBathrooms: true,
    defaultRooms: 2,
    requiredFields: ['area', 'rooms', 'bathrooms', 'floor', 'totalFloors', 'deposit', 'monthlyRent'],
    helpText: '아파트의 경우 층수 정보가 필요합니다.',
  },
  [PropertyType.VILLA]: {
    showRooms: true,
    showBathrooms: true,
    defaultRooms: 2,
    requiredFields: ['area', 'rooms', 'bathrooms', 'deposit', 'monthlyRent'],
    helpText: '빌라는 다세대 주택으로 분류됩니다.',
  },
  [PropertyType.OFFICETEL]: {
    showRooms: false,
    showBathrooms: true,
    defaultRooms: 1,
    requiredFields: ['area', 'floor', 'totalFloors', 'deposit', 'monthlyRent'],
    helpText: '오피스텔은 업무와 주거가 가능한 복합용도 건물입니다.',
  },
  [PropertyType.HOUSE]: {
    showRooms: true,
    showBathrooms: true,
    defaultRooms: 3,
    requiredFields: ['area', 'rooms', 'bathrooms', 'deposit', 'monthlyRent'],
    helpText: '단독주택은 독립된 주거공간입니다.',
  },
};

// Rental type specific configuration
export const RENTAL_TYPE_CONFIG = {
  [RentalType.MONTHLY]: {
    showDeposit: true,
    showMonthlyRent: true,
    showMaintenanceFee: true,
    requiredFields: ['deposit', 'monthlyRent'],
    helpText: '보증금과 월세를 모두 입력해주세요.',
  },
  [RentalType.JEONSE]: {
    showDeposit: true,
    showMonthlyRent: false,
    showMaintenanceFee: true,
    requiredFields: ['deposit'],
    helpText: '전세는 보증금만 입력하시면 됩니다.',
  },
  [RentalType.SALE]: {
    showDeposit: false,
    showMonthlyRent: false,
    showMaintenanceFee: true,
    requiredFields: [],
    helpText: '매매의 경우 매매가격을 보증금 필드에 입력해주세요.',
  },
};

// Common property options
export const PROPERTY_OPTIONS = [
  // 보안시설
  'CCTV',
  '출입통제시스템',
  '인터폰',
  '도어락',
  '경비실',
  
  // 편의시설
  '에어컨',
  '세탁기',
  '냉장고',
  '가스레인지',
  '전자레인지',
  '인덕션',
  '식기세척기',
  '건조기',
  '엘리베이터',
  '주차장',
  
  // 주변환경
  '지하철역 근처',
  '버스정류장 근처',
  '학교 근처',
  '병원 근처',
  '마트 근처',
  '공원 근처',
  '편의점 근처',
  
  // 기타
  '베란다',
  '발코니',
  '옥상',
  '반려동물 가능',
  '흡연 가능',
  '단기임대 가능',
];